### ISSUE TITLE
Partition WebSockets Limits to prevent side channels

### ISSUE THREAD

**Issue by pes10k** [brave/brave-browser#19990]
Currently WebSockets have a large global limit in the Chromium.  This limit is shared by all open sites.  Trackers on those sites can use this limit to communicate across site boundaries, and so track users across the web, circumventing many of Brave's partitioning protections.

To prevent the most dangerous forms of this attack (attacks that target the largest limited-but-unpartitioned pools in Chromium), Brave should do the following:

- Partition: impose a per-site (as determined by top-level frame eTLD+1) cap of how many websockets can be open.  Initial suggestion is 10, but any cap thats << the global cap would work
- De-limit: remove, or greatly expand, the global cap, to increase the number of eTLD+1s an attack has to control to carry out the attack

Once this is done for websockets, we could look for other connection pools and global limits to apply similar mitigations to.

**Comment by stephendonner** [comment]
@goodov can we get a testplan for this when you get a chance?  Thanks!

**Comment by goodov** [comment]
@pes10k wdyt about adding a test page for WebSockets limits? A manual testing can have many steps here and it's not easy to get all expectations from different frames in a quick manner.

**Comment by pes10k** [comment]
I've requested that devops create a websocket server we can use here.  That issue is here: https://github.com/brave/devops/issues/6722

Once the above issue is complete, i can implement a QA test

**Comment by pes10k** [comment]
QA test here: https://dev-pages.brave.software/network-state/web-sockets.html

**Comment by LaurenWags** [comment]
Verified using

```
Brave	1.36.94 Chromium: 98.0.4758.87 (Official Build) beta (x86_64) 
Revision	e4cd00f135fb4d8edc64c8aa6ecbe7cc79ebb3b2-refs/branch-heads/4758@{#1002}
OS	macOS Version 12.2.1 (Build 21D62)
```

Confirmed able to run the web sockets test on https://dev-pages.brave.software/network-state/web-sockets.html with default shield settings:

<img width="1201" alt="SS1" src="https://user-images.githubusercontent.com/28145373/154143703-92db4f22-8823-41d7-9302-7b14883251f5.png">

Confirmed when shields are down more than 10 connections can be created on the same page:

<img width="1194" alt="SS2" src="https://user-images.githubusercontent.com/28145373/154143722-daf27c53-d1ad-4b5b-ad0a-7f4e70acbf34.png">

Confirmed able to run the test (with shields up) on both of the following pages simultaneously:
https://dev-pages.brave.software/network-state/web-sockets.html
https://dev-pages.bravesoftware.com/network-state/web-sockets.html

<img width="2402" alt="SS3" src="https://user-images.githubusercontent.com/28145373/154143735-c3cc1103-eb37-4f35-a491-b2e7221508e2.png">

**Comment by srirambv** [comment]
Verification passed on Oppo Reno 5 with Android 12 running 1.36.105 x64 build
- Verified with shields enabled only 10 connection are created and 11th one errors out
- Verified with shields disabled more than 10 connections can be created

**`https://dev-pages.brave.software`**

Shields Enabled | Shields Disabled
-- | --
![image](https://user-images.githubusercontent.com/17010094/155700490-4ee8243e-af18-4cbf-8100-cdc1a2e3c9fa.png) | ![image](https://user-images.githubusercontent.com/17010094/155700519-2ff8a5ae-6590-4dd6-ba00-a0b9268c59dd.png)


**`https://dev-pages.bravesoftware.com`**

Shields Enabled | Shields Disabled
-- | --
![image](https://user-images.githubusercontent.com/17010094/155700443-d97152d4-58eb-4d7b-b8e0-495c6a1c8129.png) | ![image](https://user-images.githubusercontent.com/17010094/155700464-e592d7db-6397-4c0f-9fa6-15154bd0b76c.png)

----------------
Verification passed on Samsung Tab A with Android 10 running 1.36.105 x64 build
- Verified with shields enabled only 10 connection are created and 11th one errors out
- Verified with shields disabled more than 10 connections can be created

**`https://dev-pages.brave.software`**

Shields Enabled | Shields Disabled
-- | --
![image](https://user-images.githubusercontent.com/17010094/155700587-1121bc9c-4ea9-420f-9054-b0b67e5d7e42.png) | ![image](https://user-images.githubusercontent.com/17010094/155700610-14b96220-3848-46a8-b84d-2ea9a64d17cc.png)

**`https://dev-pages.bravesoftware.com`**

Shields Enabled | Shields Disabled
-- | --
![image](https://user-images.githubusercontent.com/17010094/155700551-e48ae07c-f838-4844-ae43-3769d8da3ea1.png) | ![image](https://user-images.githubusercontent.com/17010094/155700569-e3915360-f310-4fbd-a51d-3c9cff3da08e.png)

**Comment by jonathanpdunne** [comment]
I'm currently opening around ~45 web socket connections from an extension and being hampered by this limit recently. We haven’t noticed any problems with unpartitioned pools with our extension so far, so would be interested in your appetite to increase the limit - it looks like this has been considered previously. Given Chromium extensions have to pass a review process prior to publishing, an alternative might be to lift the cap for an application in the extension environment entirely. However, unpacked extensions might be an exception to this rule.

**Comment by goodov** [comment]
> I'm currently opening around ~45 web socket connections from an extension and being hampered by this limit recently. We haven’t noticed any problems with unpartitioned pools with our extension so far, so would be interested in your appetite to increase the limit - it looks like this has been considered previously. Given Chromium extensions have to pass a review process prior to publishing, an alternative might be to lift the cap for an application in the extension environment entirely. However, unpacked extensions might be an exception to this rule.

cc @pes10k would like to hear your opinion here. I believe technically we can ignore/increase the limit for extensions.

**Comment by pes10k** [comment]
Sure, extensions already have global / unpartitioned storage, so imposing the limit on extensions isn't useful.

That said, i think this (~45 parallel websockets w/in the same partition) is a pretty tail case.  If we can easily remove the limitation for extensions (say 1-2 days of work) i think its worth doing.  Otherwise, i think its a P4 or P5 kind of thing (it'd be good to address, but its low priority, given other tasks)

**Comment by goodov** [comment]
> I'm currently opening around ~45 web socket connections from an extension and being hampered by this limit recently. We haven’t noticed any problems with unpartitioned pools with our extension so far, so would be interested in your appetite to increase the limit - it looks like this has been considered previously. Given Chromium extensions have to pass a review process prior to publishing, an alternative might be to lift the cap for an application in the extension environment entirely. However, unpacked extensions might be an exception to this rule.

Should be fixed in the upcoming `nightly` build. If you can check it, would be amazing :)

**Comment by jonathanpdunne** [comment]
> @jonathanpdunne Should be fixed in the upcoming `nightly` build. If you can check it, would be amazing :)

@goodov looks like it's fixed! Big thanks for attending to this issue, really appreciate it.
